File: 백준/Gold/14500. 테트로미노/테트로미노.cc
#include <iostream>
using namespace std;

int N, M;
int arr[500][500];
int visited[500][500] = { 0 };
int dx[4] = {-1, 1, 0, 0};
int dy[4] = {0, 0, -1, 1};
int ans = 0;

void dfs(int x, int y, int cnt, int sum) {
    if (cnt == 4) {
        ans = max(ans, sum);
        return;
    }

    for (int i = 0; i < 4; i++) {
        int nx = x + dx[i];
        int ny = y + dy[i];
        if (nx >= 0 && nx < N && ny >= 0 && ny < M) {
            if (!visited[nx][ny]) {
                visited[nx][ny] =  1;
                dfs(nx, ny, cnt + 1, sum + arr[nx][ny]);
                visited[nx][ny] = 0;
            }
        }
    }
}

void T(int x, int y) {
    for (int i = 0; i < 4; i++) {
        int sum = arr[x][y];
        int validCells = 0;

        for (int dir = 0; dir < 4; dir++) {
            if (dir == i) continue;
            int nx = x + dx[dir];
            int ny = y + dy[dir];
            if (nx >= 0 && nx < N && ny >= 0 && ny < M) {
                sum += arr[nx][ny];
                validCells++;
            } 
        }
        if (validCells == 3) ans = max(ans, sum);
    }
}

int main() {
    cin >> N >> M;

    for (int i = 0; i < N; i++) {
        for (int j = 0; j < M; j++) {
            cin >> arr[i][j];
        }
    }

    for (int i = 0; i < N; i++) {
        for (int j = 0; j < M; j++) {
            visited[i][j] = 1;
            dfs(i, j, 1, arr[i][j]);
            visited[i][j] = 0;
            T(i, j);

        }
    }

    cout << ans;

    return 0;

}

File: 백준/Gold/14500. 테트로미노/테트로미노.py
n, m = map(int, input().split())
arr = [list(map(int, input().split())) for _ in range(n)]

dx = [-1, 1, 0, 0]
dy = [0, 0, -1, 1]

def dfs(x, y, cnt, sum):
    global answer
    if cnt == 3:
        answer = max(answer, sum)
        return
    for i in range(4):
        nx = x + dx[i]
        ny = y + dy[i]
        if 0 <= nx < n and 0 <= ny < m and not visited[nx][ny]:
            visited[nx][ny] = True
            dfs(nx, ny, cnt + 1, sum + arr[nx][ny])
            visited[nx][ny] = False # 백트래킹

def T_shape(x, y):
    global answer
    sum = arr[x][y]
    len = 1
    for i in range(4):
        nx = x + dx[i]
        ny = y + dy[i]
        if 0 <= nx < n and 0 <= ny < m:
            sum += arr[nx][ny]
            len += 1
    if len == 5:
        for i in range(4):
            nx = x + dx[i]
            ny = y + dy[i]
            answer = max(answer, sum - arr[nx][ny])
    elif len == 4:
        answer = max(answer, sum)

answer = 0
visited = [[False] * m for _ in range(n)]

for i in range(n):
    for j in range(m):
        visited[i][j] = True # 여기서 visited를 새로 생성하면 4중 for문으로 시간초과 뜨니까 백트래킹 처리하기!!!
        dfs(i, j, 0, arr[i][j])
        visited[i][j] = False
        T_shape(i, j)

print(answer)